libs/x-content/src/main/java/org/elasticsearch/xcontent/FlatteningXContentParser.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.xcontent;
+
+import java.io.IOException;
+
+/**
+ * A subclass of XContentSubParser that provides the functionality to flatten
+ * the field names by prefixing them with the provided parent name.
+ */
+public class FlatteningXContentParser extends XContentSubParser {
+    private final String parentName;
+    private static final char DELIMITER = '.';
+
+    /**
+     * Constructs a FlatteningXContentParser with the given parent name and wraps an existing XContentParser.
+     *
+     * @param parser The XContentParser to be wrapped and extended with flattening functionality.
+     * @param parentName The parent name to be used as a prefix for immediate children.
+     */
+    public FlatteningXContentParser(XContentParser parser, String parentName) {
+        super(parser);
+        this.parentName = parentName;
+    }
+
+    /**
+     * Retrieves the name of the current field being parsed. If the current parsing level is 1,
+     * the returned field name will be constructed by prepending the parent name to the
+     * delegate's currentFieldName, otherwise just delegate.
+     *
+     * @return The current field name, potentially modified by prepending the parent name as a prefix.
+     * @throws IOException If an I/O error occurs during parsing.
+     */
+    @Override
+    public String currentName() throws IOException {
```

Comment:
I find it a bit surprising that the only method that needs to be overridden here is `currentName`. If we are flattening docs, shouldn't nextToken get adjusted too in some ways? I find this a little hard to follow.

Reply:
Because we need to distinguish between deeply nested leaf fields, and dynamically created mappers that accept an object, we only flatten a single level at a time.  We don't need to adjust anything apart from the lowest-level name, because once we move into a new object we will either pass it to a mapper to parse, or wrap it again.

Reply:
Here we are using the logic of `XContentSubParser` that I extend. There is keeping track of the current nested level.

Reply:
We still expect nextToken to be called, but it delegates to the XContentSubParser implementation (which keeps track of levels, etc).  But yes, +1 to making the parser private.

Reply:
I agree on the private parser, but as @romseygeek said we must not throw when nextToken is called, we expect to call the parent's class nextToken.

Reply:
sounds good, thanks.Let's also see if there's other assumptions we are making that could be enforced directly in the parser to prevent it from being misused. Maybe nothing to do, not exactly sure.